File: PythonRewrite/assignments/musicians/band_supervisor.py
import time
import typing
from multiprocessing import Pipe, Process, Queue
from multiprocessing.connection import Connection

from assignments.musicians.musician import Musician
from assignments.musicians.skill_level import SkillLevel


delay: typing.Final[float] = 0.75


def start_band(max_retries: int):
    queue = Queue(0)
    members = dict()
    members['singer'] = add_band_member('singer', SkillLevel.GOOD, queue)
    members['bass'] = add_band_member('bass', SkillLevel.GOOD, queue)
    members['drum'] = add_band_member('drum', SkillLevel.BAD, queue)
    members['guitar'] = add_band_member('guitar', SkillLevel.GOOD, queue)
    while True:
        role, skill_level = queue.get(block=True)
        max_retries -= 1
        members[role] = None
        if max_retries <= 0:
            print('The manager is mad and fired the whole band!')
            for conn in members.values():
                if conn is not None:
                    conn.send(0)
            time.sleep(3.0)
            break

        members[role] = add_band_member(role, skill_level, queue)


def add_band_member(role: str, skill_level: SkillLevel, queue: Queue) -> Connection:
    receiver, sender = Pipe(duplex=False)
    musician = Musician(role, skill_level, receiver)
    process = Process(target=loop, args=(musician, queue,), daemon=True)
    process.start()
    return sender


def loop(m: Musician, q: Queue):
    while True:
        time.sleep(delay)
        play_result = m.play_sound()
        if play_result:
            continue

        q.put((m.role, m.skill_level))
        break

File: Cpp/Musicians/BandSupervisor.cpp
//
// Created by 周哲輝 on 2022/03/28.
//

#include "BandSupervisor.h"
#include <chrono>
#include <iostream>
#include "Musician.h"

std::mutex BandSupervisor::mutex = std::mutex();
std::vector<std::future<void>> BandSupervisor::handles = std::vector<std::future<void>>();

void BandSupervisor::add_band_member(const std::string &role, const SkillLevel &skillLevel, std::deque<std::tuple<std::string, SkillLevel>>& queue, std::unordered_map<std::string, std::atomic_bool*>& members) noexcept {
    auto future = std::async(std::launch::async, [&]() {
        auto musician = Musician(role, skillLevel, members);
        while (true) {
            std::this_thread::sleep_for(std::chrono::milliseconds(delay));
            auto play_result = musician.play_sound();
            if (play_result) {
                continue;
            }

            {
                auto lock = std::lock_guard(mutex);
                queue.emplace_back(std::tuple<std::string, SkillLevel>(role, skillLevel));
            }
            break;
        }
    });
    handles.emplace_back(std::move(future));
}

void BandSupervisor::start_band(int maxRetries) noexcept {
    auto queue = std::deque<std::tuple<std::string, SkillLevel>>();
    auto members = std::unordered_map<std::string, std::atomic_bool*>();
    add_band_member("singer", SkillLevel::GOOD, queue, members);
    add_band_member("bass", SkillLevel::GOOD, queue, members);
    add_band_member("drum", SkillLevel::BAD, queue, members);
    add_band_member("guitar", SkillLevel::GOOD, queue, members);

    do {
        if (queue.empty()) {
            continue;
        }

        auto [role, skill_level] = queue.back();
        queue.pop_back();
        members.erase(role);
        --maxRetries;
        if (maxRetries <= 0) {
            std::cout << "The manager is mad and fired the whole band!\n";
            for (auto& [k, v] : members) {
                v->store(true);
            }
            std::this_thread::sleep_for(std::chrono::seconds(3));
            break;
        }

        add_band_member(role, skill_level, queue, members);
    } while (maxRetries > 0);
}
